lib/tesla/middleware/telemetry.ex

```diff
@@ -13,21 +13,102 @@ if Code.ensure_loaded?(:telemetry) do
 
     end
 
-    :telemetry.attach("my-tesla-telemetry", [:tesla, :request], fn event, time, meta, config ->
+    :telemetry.attach("my-tesla-telemetry", [:tesla, :request, :stop], fn event, measurements, meta, config ->
       # Do something with the event
     end)
     ```
 
+    ## Options
+
+    * `:event_prefix` - a list of atoms to prefix to the telemetry event name. This can be set if you need to distinguish events from different clients. Defaults to `[]`
+
+    ## Telemetry Events
+
+    * `[:tesla, :request, :start]` - emitted at the beginning of the request.
+      * Measurement: `%{time: System.monotonic_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :stop]` - emitted at the end of the request.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :error]` - emitted when there is an error.
+      * Measurement: `%{value: 1}`
+      * Metadata: `%{env: Tesla.Env.t, kind: Exception.kind | nil, reason: term, stacktrace: Exception.stacktrace}`
```

Comment:
Hrm, should we make a distinction between error tuples and exceptions? It feels weird to put them in the same bucket but I am not sure which approach is effectively better. Which kind of errors do we have in the error tuple? Connection errors? What about a 500 response? Is that still an OK tuple?

Reply:
I had an exception event in the [initial design](teamon/tesla/pull/324/commits/c1e01c1a13a9ab9a982cdfd00c69cce16b6c14c9) to distinguish but prior art in tracing seems to favor lumping them together. My initial inclination was the same as yours.

https://opentracing.io/docs/overview/tags-logs-baggage/
https://opentracing.io/specification/conventions/

@tsloughter may have a better handle on if these are being carried into OTel (I'm guessing so for backward compatibility).

>Which kind of errors do we have in the error tuple? Connection errors? What about a 500 response? Is that still an OK tuple?

That is a great question. API designers are going to handle these differently. @teamon can correct me if I'm wrong but, in Tesla's case, an ok tuple is returned for any request which completes and all error tuples are the result of a failure to complete the request, including timeouts, adapter failures, DNS lookup failures, etc.

One issue I'm seeing with establishing a good pattern is passing the result to stop as metadata. If we try to do this in the after, we don't have a way to access that result. This is probably OK in some cases, but I've found that the metadata is frequently used in our metrics in the wild. I'm struggling to figure out a pattern that doesn't require dropping the after clause and writing out more granular events for each possible situation.